### PR TITLE
Fix spring kafka context leak when batch listener is retried

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/10698
In later versions of spring kafka `BatchInterceptor` does not work correctly for retries. On retry only the `intercept` method is called that starts the span, but `success` and `failure` that end it are not called. This results in a context leak. This PR implements javaagent instrumentation for batch listeners without using `BatchInterceptor` and modifies the `BatchInterceptor` in library instrumentation to ignore retries. Tests are altered to also test retries.

### CODE REVIEW

File: instrumentation/spring/spring-kafka-2.7/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/v2_7/SpringKafkaTest.java [hunk 1]
```diff
@@ -358,44 +407,79 @@ void shouldHandleFailureInKafkaBatchListener() {
...
-                            satisfies(
-                                SemanticAttributes.MESSAGING_CLIENT_ID,
-                                stringAssert -> stringAssert.startsWith("consumer")),
-                            equalTo(SemanticAttributes.MESSAGING_BATCH_MESSAGE_COUNT, 1)),
-                span -> span.hasName("consumer").hasParent(trace.getSpan(1))));
+        });
+
+    if (Boolean.getBoolean("testLatestDeps")) {
+      // latest dep tests call receive once and only retry the failed process step
+      assertions.add(
```
Comment (zeitlinger) [inline instrumentation/spring/spring-kafka-2.7/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/v2_7/SpringKafkaTest.java:414]: this test is a bit hard to follow - maybe extract some common assertions

**Review by zeitlinger — APPROVED** [review]

File: instrumentation/spring/spring-kafka-2.7/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/v2_7/SpringKafkaTest.java (hunk 1 above)
Comment (laurit) [reply instrumentation/spring/spring-kafka-2.7/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/v2_7/SpringKafkaTest.java:389]: extracted common assertions

Comment (zeitlinger) [reply instrumentation/spring/spring-kafka-2.7/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/v2_7/SpringKafkaTest.java:389]: great - much easier to read :smile:

**Review by zeitlinger — APPROVED** [review]

File: instrumentation/spring/spring-kafka-2.7/testing/src/main/java/io/opentelemetry/testing/AbstractSpringKafkaNoReceiveTelemetryTest.java [hunk 2]
```diff
@@ -293,19 +321,40 @@ void shouldHandleFailureInKafkaBatchListener() {
...
-                                    SemanticAttributes.MESSAGING_CLIENT_ID,
-                                    stringAssert -> stringAssert.startsWith("consumer")),
-                                equalTo(SemanticAttributes.MESSAGING_BATCH_MESSAGE_COUNT, 1)),
-                    span -> span.hasName("consumer").hasParent(trace.getSpan(0))));
+                            .hasAttributesSatisfyingExactly(processAttributes),
+                    span -> span.hasName("consumer").hasParent(trace.getSpan(0))),
+            trace -> {
+              if (Boolean.getBoolean("testLatestDeps")) {
+                // in latest dep tests process spans are not created for retries because spring does
+                // not call the success/failure methods on the BatchInterceptor for reties
```
Comment (trask) [inline instrumentation/spring/spring-kafka-2.7/testing/src/main/java/io/opentelemetry/testing/AbstractSpringKafkaNoReceiveTelemetryTest.java:329]: is this comment relevant for the javaagent instrumentation?

Comment (laurit) [reply instrumentation/spring/spring-kafka-2.7/testing/src/main/java/io/opentelemetry/testing/AbstractSpringKafkaNoReceiveTelemetryTest.java:331]: Thanks I left the comment as it is but change the condition so that this branch is only used for library instrumentation. I think ideally we should implement the library instrumentation in a different way to have the library instrumentation create correct traces for retried batches.
